### ISSUE TITLE
[Baselines] ML Inference for Photo-Z/Physical Properties

### ISSUE THREAD

**Issue by lhparker1** [AstroPile/FlatironMeeting2024#20]
# ML Inference for Photo-Z/Physical Properties

Building inference baselines for Photo-Z and Physical Property Prediction. 

Contacts: Liam Parker, Kartheik Iyer
Participants: Liam Parker, Kartheik Iyer

## Goals and deliverable
Build two levels to the inference pipeline:

1. Straight-up prediction in-distribution
2. Generalization out-of-distribution (the real science problem)

Can implement a couple of different models for this, all CNN + NF (can also do it from simple photometry as a baseline too).

## Resources needed
Mainly need ML background
  
## Detailed description
See goals and deliverable

**Comment by lhparker1** [comment]
We want to provide a couple of different levels at which people can perform this task, ranging from "really low bar" and not scientifically useful to "medium bar" and somewhat scientifically useful. We'll detail these below:

**Level 0**: Random train-test split, no eval on ood data
**Level 1**: Impose mag-z/high-redshift/effective size cut and eval ood performance 

Also might want to include some photometry baseline (with either an MLP and an easy mlp fitter)

**Comment by lhparker1** [comment]
The basic training class for the CNN is set up now, and is producing this. Will train overnight and check on the morning. 
<img width="1354" alt="Screenshot 2024-03-26 at 4 22 04 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/78778a4a-5ec4-40d1-80ee-e8787689cf74">

**Comment by lhparker1** [comment]
Built a photo-z wrapper that takes a train_dataset and a test_dataset as well as the feature_flag and label_flag, and gives you back x and y to be used directly for photo_z estimation. The idea is to do splitting beforehand and model training either with the provided dataloaders or directly in a pytorch lightning environ. 
<img width="1265" alt="Screenshot 2024-03-27 at 5 05 45 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/e639be56-41c9-4cb7-b953-6bbb985b2452">

**Comment by lhparker1** [comment]
Once you've initialized it, you just need to call `prepare_data()` on it to get all of the statistics computed and the collate functions / etc. all set up.
<img width="1267" alt="Screenshot 2024-03-27 at 5 08 24 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/21d4352c-6262-45e1-a6e7-5e23f178ca7f">

**Comment by lhparker1** [comment]
you can then just get the dataloaders from here
<img width="439" alt="Screenshot 2024-03-27 at 5 10 20 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/eb10f2cf-4c17-4122-9bb5-bf241347dad3">

**Comment by lhparker1** [comment]
I have a basic CNN class set up and the model can now train pretty easily with just two lines of code
<img width="1339" alt="Screenshot 2024-03-28 at 5 17 34 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/9e5c025d-ee69-4ea0-8dae-33ce91262645">

**Comment by lhparker1** [comment]
Training on cross-matched HSC and SDSS - but should be able to do with any cross-matched image-spectrum set

<img width="1345" alt="Screenshot 2024-03-28 at 5 17 54 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/ad7335e8-c940-4ce5-938c-4c6feea47a4f">

**Comment by lhparker1** [comment]
It's only 1000 epochs and I trained it pretty hand-wavy to get R^2 of 0.6 (I know I should be able to get around 0.76 on DECaLS + DESI) 
<img width="1366" alt="Screenshot 2024-03-28 at 4 47 07 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/5d74b944-5c0a-4eac-b831-a4287da3010c">

**Comment by lhparker1** [comment]
Training for some reason isn't working (the pytorch model runs and says it trains but the loss doesn't move) - will fix bug tomorrow.

**Comment by lhparker1** [comment]
Never mind, the normalizing flow trains now too 
<img width="1346" alt="Screenshot 2024-03-28 at 5 27 15 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/ffc8d065-2677-47bf-b49d-c06156b2c6e2">

**Comment by lhparker1** [comment]
examples samples 
<img width="533" alt="Screenshot 2024-03-28 at 5 38 44 PM" src="https://github.com/AstroPile/FlatironMeeting2024/assets/86175266/c7b6c058-2b64-4b86-a571-ff5df03bc758">
